### PR TITLE
Generic interface for Mesh

### PR DESCRIPTION
**Pull request by stefanoborini**
This PR introduces the generic interface for datasets and applies it to the Mesh object, with minimal disruption to the internals and no disruptions to the old interface. 
We expect a loss of efficiency but that's the price to pay for not knowing the uid->type association. At the moment, it's probably not wise to implement optimization strategies to solve this issue, and will be part of a separate PR once we migrated all objects.

### CODE REVIEW

File: simphony/cuds/abc_dataset.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            The object item.
+
+        Raises
+        ------
+        KeyError :
+            if any of the ids passed as parameters are not in the dataset.
+        """
+
+    @abstractmethod
+    def has(self, uid):  # pragma: no cover
```
Comment (mehdisadeghi) [inline simphony/cuds/abc_dataset.py]: I suggest to implement `__contains__` instead of this.
https://docs.python.org/3/reference/datamodel.html#object.__contains__

File: simphony/cuds/abc_dataset.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        item_type : CUDSItem
+            The CUDSItem enum of the type of the items to return the count of.
+
+        Returns
+        -------
+        True if the type is present, False otherwise.
+        """
+
+    @abstractmethod
+    def count_of(self, item_type):  # pragma: no cover
```
Comment (mehdisadeghi) [inline simphony/cuds/abc_dataset.py]: Shell we implement `__len__` beside this? Of course it will return the whole number of elements inside the dataset.

File: simphony/cuds/abc_dataset.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
+# -*- coding: utf-8 -*-
+from abc import ABCMeta, abstractmethod
+
+
+class ABCDataset(object):
```
Comment (mehdisadeghi) [inline simphony/cuds/abc_dataset.py]: DataSets must conform the the structore of `CUBA.DATA_SET`, i.e. they need `uid`, `name`, etc defined.

File: simphony/cuds/abc_mesh.py [hunk 4]
```diff
@@ -1,21 +1,254 @@
...
+            the uids of the objects to be removed.
+
+        Raises
+        ------
+        KeyError :
+            If any object doesn't exist.
+        """
+        raise NotImplementedError("Remove is not implemented for Mesh")
+
+    def iter(self, uids=None, item_type=None):
```
Comment (mehdisadeghi) [inline simphony/cuds/abc_mesh.py]: Docstring for `item_type` is missing

File: simphony/cuds/abc_dataset.py (hunk 1 above)
Comment (mehdisadeghi) [inline simphony/cuds/abc_dataset.py]: However, the same applies to all the other pythonic container methods. Perhaps it's better to leave this the subject of another PR. (And I see we have `has_something` methods there)

File: simphony/cuds/abc_dataset.py (hunk 3 above)
Comment (stefanoborini) [inline simphony/cuds/abc_dataset.py]: Ok then I probably used the wrong name. I just picked dataset but I am not sure about it.
What is the generic name for things like particles, mesh, and lattice?

File: simphony/cuds/abc_dataset.py (hunk 1 above)
Comment (stefanoborini) [inline simphony/cuds/abc_dataset.py]: well, if you do so, it's certainly possible, but at that point we should just reimplement the dict() interface... :)

File: simphony/cuds/abc_dataset.py (hunk 2 above)
Comment (stefanoborini) [inline simphony/cuds/abc_dataset.py]: That's possible, yes.

File: simphony/cuds/abc_dataset.py (hunk 3 above)
Comment (mehdisadeghi) [inline simphony/cuds/abc_dataset.py]: DataSet!

**Comment by stefanoborini** [comment]
@mehdisadeghi please re-review.

File: simphony/cuds/abc_dataset.py (hunk 1 above)
Comment (stefanoborini) [inline simphony/cuds/abc_dataset.py]: Added.

File: simphony/cuds/abc_dataset.py (hunk 2 above)
Comment (stefanoborini) [inline simphony/cuds/abc_dataset.py]: Added

File: simphony/cuds/abc_mesh.py (hunk 4 above)
Comment (stefanoborini) [inline simphony/cuds/abc_mesh.py]: added

File: simphony/cuds/abc_dataset.py (hunk 3 above)
Comment (mehdisadeghi) [inline simphony/cuds/abc_dataset.py]: I think the name is fine. How about adding `meta.DataSet` as a parent to get the properties?

**Comment by stefanoborini** [comment]
Some failures are due to the fact that we are now dependent on the DataSet. Some assumptions of the test are no longer valid, but they will be restored when we merge #335

I'll merge that branch into this one, but this mean that #335 must be merged first.

**Comment by stefanoborini** [comment]
Another source of problem is that with the new hierarchy involving the generated base meta classes, the data object now contains None entries, that are converted to the string "None" when written on the hdf5 file with pytables.

**Comment by stefanoborini** [comment]
The problem of None should be solved by setting appropriate defaults to strings, as from PR #343. In any case, the base classing will be done on a separate PR.
